src/utils/utils.jl

```diff
@@ -115,57 +116,77 @@ function validate_exported_names(mod::Module)
     return is_valid
 end
 
+# Make match_fn optional
+compare_values(x, y; kwargs...) = compare_values(nothing, x, y; kwargs...)
+
+# Get the default match_fn if necessary. Only call when you know you're done recursing
+_fetch_match_fn(match_fn::Function) = match_fn
+_fetch_match_fn(::Nothing) = isequivalent
+
+# Whether to stop recursing and apply the match_fn
+_is_compare_directly(::DataType, ::DataType) = true
+_is_compare_directly(::T, ::U) where {T, U} = true
+_is_compare_directly(::T, ::T) where {T} = isempty(fieldnames(T))
+
 """
 Recursively compares struct values. Prints all mismatched values to stdout.
 
 # Arguments
-
+  - `match_fn`: optional, a function used to determine whether two values match in the base
+    case of the recursion. If `nothing` or not specified, the default implementation uses
+    `IS.isequivalent`.
   - `x::T`: First value
-  - `y::T`: Second value
+  - `y::U`: Second value
   - `compare_uuids::Bool = false`: Compare any UUID in the object or composed objects.
   - `exclude::Set{Symbol} = Set{Symbol}(): Fields to exclude from comparison. Passed on
      recursively and so applied per type.
 """
-function compare_values(
-    x::T,
-    y::T;
-    compare_uuids = false,
-    exclude = Set{Symbol}(),
-) where {T}
+function compare_values(match_fn::Union{Function, Nothing}, x::T, y::U;
```

Comment:
@GabrielKS In what cases do we pass different types to this function? How can two values that are of different types be considered equivalent?

Reply:
We want for instance `isequivalent(0, 0.0)`, both because it makes sense and for backwards compatibility. In a broader sense we want to impose minimal constraints on how `match_fn` works, if it wants to specify that things of different type don't match then it is free to do so.

Reply:
Isn't the purpose of `compare_values` to `"Recursively compares struct values."`? We should always be comparing identical types. Are we ever calling it to test integers and floats?